### ISSUE TITLE
NVMExpress Shutdown on EfiResetSystem (Bugzilla Bug 224)

### ISSUE THREAD

**Issue by tianocore-issues** [tianocore/edk2#6732]
This issue was created automatically with [bugzilla2github](https://github.com/makubacki/bugzilla2github/tree/bz2gh_tianocore)
# Bugzilla Bug 224
Date: 2016-11-11T20:12:53+00:00
From: Michael Turner &lt;<Michael.Turner>&gt;
To:   @niruiyu
CC:   feng.tian, @lersek, yonghong.zhu

Last updated: 2017-09-05T22:59:19+00:00

**Comment by tianocore-issues** [comment]
## Comment 889
Date: 2016-11-11 20:12:53 +0000
From: Michael Turner &lt;<Michael.Turner>&gt;

- Industry Specification: ---
- Releases to Fix: EDK II Master
- Target OS: ---
- Bugzilla Assignee(s): @niruiyu

Created attachment 20
git diff output

NVMExpress drives require advanced notification prior to a reset to prevent data corruption.

Attachment: [BZ-20-changes.txt](https://github.com/tianocore/user-attachments/raw/refs/heads/main/tianocore/edk2/BZ-20-changes.txt)

**Comment by tianocore-issues** [comment]
## Comment 910
Date: 2016-11-14 04:17:08 +0000
From: @lersek

(In reply to Michael Turner from comment #0)
&gt; Created attachment 20 [details]
&gt; git diff output
&gt; 
&gt; NVMExpress drives require advanced notification prior to a reset to prevent
&gt; data corruption.

Michael,

can you please send the patch to the list for review?

(Also, the name &quot;gEfiResetNotifyCallbackProtocolGuid&quot; seems to imply a standard protocol, but I don&apos;t see that name anywhere in edk2 or in the spec. Is there an in-progress Mantis ticket perhaps?)

**Comment by tianocore-issues** [comment]
## Comment 936
Date: 2016-11-15 03:55:23 +0000
From: Feng Tian &lt;<feng.tian>&gt;

Mike &amp; Laszlo,

I think this patch is used to execute steps defined in NVMe spec (see NVMe1.2 chapter 7.6.2) when a power-off condition happens.

My opinion is it should be a platform behavior rather than core behavior, which means platform driver should hook shutdown(S3/S4) to perform NVMe controller shutdown.

This way could avoid introducing a implementation specific guid.

**Comment by tianocore-issues** [comment]
## Comment 940
Date: 2016-11-15 09:22:06 +0000
From: @lersek

Generally speaking, how can a platform hook shutdown? Do you mean the ResetSystemLib interfaces that are called from MdeModulePkg/Universal/ResetSystemRuntimeDxe (which provides gEfiResetArchProtocolGuid)? Thanks.

**Comment by tianocore-issues** [comment]
## Comment 947
Date: 2016-11-15 17:26:24 +0000
From: Michael Turner &lt;<Michael.Turner>&gt;

I do apologize for exposing our private reset notification.

However, I do not agree that this problem is a platform issue.  This is a problem with the UEFI Core NVM/E driver (and the TPM driver, for that matter) that have been implemented without regard to certain device specifications.

**Comment by tianocore-issues** [comment]
## Comment 949
Date: 2016-11-15 21:15:59 +0000
From: Feng Tian &lt;<feng.tian>&gt;

Mike,

What I mean is although the device shutdown notification is defined in NVMe spec we can have a platform-specific solution to solve this.

for example, disconnect NVMe controller and set CC.SHN bit at platform code for normal shutdown.

Another point of view is NVMe is the only device I know which needs a shutdown notification. so whether introducing such guid is valuable may depend on how device shutdown notification becomes popular.

PS: do you know what&apos;s the impact of not executing the NVMe shutdown process? NVMe spec doesn&apos;t say it...

Laszlo,

I also don&apos;t know how MS implements/signals this gEfiResetNotifyCallbackProtocolGuid. but I guess so, that&apos;s install this protocol guid in ResetSystemLib. Mike, please correct me if I am wrong.

**Comment by tianocore-issues** [comment]
## Comment 959
Date: 2016-11-16 11:14:14 +0000
From: @lersek

(In reply to Feng Tian from comment #5)

&gt; Laszlo,
&gt; 
&gt; I also don&apos;t know how MS implements/signals this
&gt; gEfiResetNotifyCallbackProtocolGuid. but I guess so, that&apos;s install this
&gt; protocol guid in ResetSystemLib. Mike, please correct me if I am wrong.

My question is how you *suggest* to implement the shutdown / reset hook.

In other words, how a platform *should* add its cleanup code, so that it runs when the ResetSystem() runtime service is called. Thanks!

**Comment by tianocore-issues** [comment]
## Comment 974
Date: 2016-11-17 12:53:46 +0000
From: Michael Turner &lt;<Michael.Turner>&gt;

We have made public our ResetSystem interface that we have used over the last year.

https://github.com/Microsoft/MS_UEFI/tree/share/ResetSystem

This is used by both our NVME code and TPM code (yes, the TPM requires a shutdown notification also - just as a counter to &quot;no other hardware needs a reset notification&quot;).

**Comment by tianocore-issues** [comment]
## Comment 977
Date: 2016-11-17 13:50:17 +0000
From: @lersek

Can we perhaps standardize this feature? It could see other uses as well; a writeback-caching filesystem implementation, for example. (The write cache would have to be flushed at reset / shutdown / ready-to-boot / legacy-boot, I assume.)

I have no current proposal for a caching filesystem, mind you, I just recall thinking about it a few months (?) back, and telling myself it would be unsafe after all because there was no shutdown / reset notification.

**Comment by tianocore-issues** [comment]
## Comment 2329
Date: 2017-06-22 20:14:24 +0000
From: Yonghong Zhu &lt;<yonghong.zhu>&gt;

Bug scrub: Assign to module owner

**Comment by tianocore-issues** [comment]
## Comment 2686
Date: 2017-09-05 22:59:19 +0000
From: @niruiyu

pushed @ 7111e46fdebca0e03f6ddfcb029522c2fe68c74b.
